### ISSUE TITLE
False-positive "cycle detected" in `const fn` with RPIT

### ISSUE THREAD

**Issue by yvt** [rust-lang/rust#99793]
I tried this code:

```rust
const fn f() -> impl Eq { g() }
const fn g() {}
// OK: const fn f() -> impl Eq { g(); }
// OK: const fn f() -> impl Eq {}
// OK: const fn f() -> impl Eq { (g(),).0 }
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a29f964e865804f8732e5d53731b3df4))

I expected to see this happen: Successful compilation

Instead, this happened:

```
error[E0391]: cycle detected when computing type of `f::{opaque#0}`
 --> src/lib.rs:1:17
  |
1 | const fn f() -> impl Eq { g() }
  |                 ^^^^^^^
  |
note: ...which requires borrow-checking `f`...
 --> src/lib.rs:1:1
  |
1 | const fn f() -> impl Eq { g() }
  | ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires processing `f`...
 --> src/lib.rs:1:1
  |
1 | const fn f() -> impl Eq { g() }
  | ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires const checking `f`...
 --> src/lib.rs:1:1
  |
1 | const fn f() -> impl Eq { g() }
  | ^^^^^^^^^^^^^^^^^^^^^^^
  = note: ...which requires computing whether `impl core::cmp::Eq` is freeze...
  = note: ...which requires evaluating trait selection obligation `impl core::cmp::Eq: core::marker::Freeze`...
  = note: ...which again requires computing type of `f::{opaque#0}`, completing the cycle
note: cycle used when checking item types in top-level module
 --> src/lib.rs:1:1
  |
1 | const fn f() -> impl Eq { g() }
  | ^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0391`.
```

### Meta

`rustc --version --verbose`:
```
rustc 1.62.1 (e092d0b6b 2022-07-16)
binary: rustc
commit-hash: e092d0b6b43f2de967af0887873151bb1c0b18d3
commit-date: 2022-07-16
host: x86_64-unknown-linux-gnu
release: 1.62.1
LLVM version: 14.0.5
```

```
rustc 1.64.0-nightly (6dbae3ad1 2022-07-25)
binary: rustc
commit-hash: 6dbae3ad19309bb541d9e76638e6aa4b5449f29a
commit-date: 2022-07-25
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by y86-dev** [comment]
I also encountered this issue on
```text
rustc 1.65.0-nightly (17cbdfd07 2022-09-13)
binary: rustc
commit-hash: 17cbdfd07178349d0a3cecb8e7dde8f915666ced
commit-date: 2022-09-13
host: x86_64-unknown-linux-gnu
release: 1.65.0-nightly
LLVM version: 15.0.0
```
The following code also works (seems like a general workaround):
```rust
const fn f() -> impl Eq { let g = g(); g }
const fn g() {}
```
